### PR TITLE
ENH: add LTI class hierarchy and lti/dlti related functions

### PR DESCRIPTION
**Pull request by ev-br**
cross-ref https://github.com/cupy/cupy/issues/7403

- [x] StateSpace
- [x] TransferFunction
- [x] ZerosPolesGain
.
- [x] dlti
- [x] dlsim
- [x] dimpulse
- [x] dstep
- [x] dbode
- [x] dfreqresp
.
- [x] lti
- [x] lsim
- [x] impulse
- [x] step 
- [x] freqresp
- [x] bode
.
- [x] cont2discrete

Also add the matrix exponential, `cupyx.scipy.linalg.expm` (needed for lsim, cont2discrete).

### CODE REVIEW

**Comment by ev-br** [comment]
DLTI hierarchy and related functions are done here. The PR is still in a draft status, for two reasons:

- it depends on gh-7652, which brings low-level conversions between formats
- static checkers complain about missing continuous LTI functions. These two, continuous and discrete LTI hierarchies, are sufficiently intertwined, so probably best to add both in the same PR. However continuous LTI evolution, `lsim`, relies on a matrix exponential, so this would require porting `scipy.linalg.expm` to CuPy.

**Comment by takagi** [comment]
We have a draft expm PR #5688, but don't hear from its author for so long, though.

**Comment by ev-br** [comment]
Update: 

- Added `expm`. For the moment, only implement the Higham'2005 algorithm [1], in a simplified form:
     -  [13/13] Pade approximant only (no selection of 3-5-7-9 orders)
     - no matrix balancing (this requires LAPACK's `*gebal` which is not available on CUDA): balancing could improve numerical stability.
     - the implementation is pure python + matmuls.

- Added `lsim`, which is essentially `expm(t*A)` for `t` a linspace array. Like in scipy.signal, the computation is a python loop with a repeated application of `expm(dt*A)`. While performance is not terrible, it can probably be improved by using an Al-Mohy & Higham 2010 algorithm for a `t` array [2]. Keeping it simple for now.

- Added `impulse` and `step`, both are straighforward applications of `lsim`. Their APIs are annoyingly polymorphic (from 2000s), am keeping them compatible with scipy.signal. One annoying issue is that both use `eigvals` to estimate a default time span given the system matrix. Am delegating to `np.linalg.eigvals`. An alternative is to estimate the boundaries for the eigenvalues using Gershgorin theorems, but that would make the default behavior incompatible with scipy.signal.

[1] https://epubs.siam.org/doi/10.1137/04061101X
[2] https://eprints.maths.manchester.ac.uk/1300/1/

**Comment by ev-br** [comment]
`<rant mode on>`  Have to admit I'm baffled by mypy complaining about ways to format module-level lists. What on earth is wrong with 

```
b = [0] * 14
b[0] = 1
b[1] = 2
...
```
`<rant mode off>`

Anyway, this is fixed in https://github.com/cupy/cupy/pull/7660/commits/c4004004f7d1300cde3e0ed5bc2c0c72ee82b052.

**Comment by ev-br** [comment]
OK, this now implements both continuous and discrete LTI functionality. Is ready from my side, modulo the need to rebase when   gh-7652 gets in.

**Comment by ev-br** [comment]
Rebased on main, so this PR is ready for review from my side.

File: cupyx/scipy/linalg/_matfuncs.py
Comment (takagi) [inline cupyx/scipy/linalg/_matfuncs.py:99]: Transferring the data of `nrmA` to CPU first would reduce the number of sync. I suppose, ideally, we could eliminate the sync to do these scaling and squaring later with keeping `s` on GPU, but it would include troublesome cuBLAS device API calls and HIP compatibility issues. It would be too much here.

Comment (takagi) [reply cupyx/scipy/linalg/_matfuncs.py:99]: Also, can I ask you to have `expm` in another PR to separate from this one?

Comment (ev-br) [reply cupyx/scipy/linalg/_matfuncs.py:99]: Would it help to keep `b` and scalars s and th13 on the device?

Comment (takagi) [reply cupyx/scipy/linalg/_matfuncs.py:99]: I suppose we need `s` on the host to the squaring:
```py
for _ in range(s):
    x = x @ x
```
We could do this repetitive matrix multiplication if we have `s` on the device, but it would require cuBLAS device API calls in a RawKernel.

Comment (ev-br) [reply cupyx/scipy/linalg/_matfuncs.py:99]: I see. Just confirm my understanding: changing the matmul into `cupy.matmul(x, x, out=x) ` also incurs synchronizarion, correct?

Comment (takagi) [reply cupyx/scipy/linalg/_matfuncs.py:99]: Both of the matmul and `cupy.matmul` do not have synchronization. In the current code, `if scale` where `scale` is evaluated as if `bool(scale)` and `float(nrmA)` have.

**Comment by ev-br** [comment]
> have expm in another PR to separate from this one?

Done in https://github.com/cupy/cupy/pull/7744.

Meanwhile I hopefully fixed the CI errors in this PR, and will rebase this PR to remove expm-related commits when gh-7744 is decided on.

**Comment by takagi** [comment]
> Done in https://github.com/cupy/cupy/pull/7744.

Thanks!

**Review by takagi — APPROVED** [review]
LGTM! I'll kick the CI again and then merge this once #7744 is merged.

**Comment by ev-br** [comment]
Rebased on main.

**Comment by ev-br** [comment]
Pushed an update to relax the contiguity checks which were failing on CI for some configurations.

**Review by takagi — APPROVED** [review]

**Comment by takagi** [comment]
Thanks, @ev-br!

**Pull request merged**
